CodeEdit: Improve render time by 2x

rune-scape opened the pull request:
with gdscript syntax highlighting:
before: 30.0385588566243ms avg draw time (1089 samples)
after: 13.0933055755083ms avg draw time (1089 samples)

without syntax highlighting:
before: 17.9521260292186ms avg draw time (1089 samples)
after: 12.9270402524754ms avg draw time (1089 samples)

#### many improvements to do with:

avoiding redraw:
- use `set_process_internal(true)` instead of queue_redraw to avoid drawing 2-3 times in the same frame when `gui_input()` is called
  - tests needed to be updated bc of this
- `set_code_hint` and `set_code_hint_draw_below` avoid redrawing when the values dont change

improving render speed:
- cache line syntax highlighting
- line numbers:
  - caching line number shaped text
  - simple enough to use TextServer RIDs directy instead of `TextLine`
  - only render line numbers on screen
- `_main_gutter_draw_callback`: use `get_hovered_gutter()` instead of multiple calls to the more expensive `get_local_mouse_pos()`
- cache `total_visible_line_count` and `Line::line_count`
- delay calculating text max width and height in case its not needed
- `GutterInfo`, which contains a custom callable in both of godots script editor gutters, was getting copied once per gutter per line per frame,  

some less impactful (more code style) things ..
- avoid creating HashMaps every draw frame, just use vectors
- const-ify copies of Vectors to avoid possible copy-on-write when theres no write

its obvious that most of the problem was the line syntax highlighting being called every frame
but whats not obvious is that searching through the dictionary and converting to and from variants was a very significant part of it

the performance improvements are especially noticeable on debug builds, as there arent any 'double draws' making the editor lag when scrolling

benchmark: [CodeEditText.zip](https://github.com/user-attachments/files/15737660/CodeEditText.zip)
tested with godot compiled using: `production=yes optimize=size tests=yes deprecated=no scu_build=yes`
bc it uses the GDScriptSyntaxHighlighter, to benchmark you have open the project in the editor

KoBeWi commented:
> use set_process_internal(true) instead of queue_redraw to avoid drawing 2-3 times in the same frame

Drawing is already limited to once per frame. It's called `queue_redraw()`, because it queues one redraw no matter how many times you call it. A different result would be a bug.

rune-scape commented:
it very much does draw multiple times per process loop (maybe different meaning than frame) bc queue_redraw adds a message to the MessageQueue which gets flushed multiple times per iteration

KoBeWi commented:
How
https://github.com/godotengine/godot/blob/e96ad5af98547df71b50c4c4695ac348638113e0/scene/main/canvas_item.cpp#L427-L433

AThousandShips commented:
How does `queue_redraw` get called multiple times per frame though? The queue gets flushed a few times but it's in different phases of the frame, and unless some code calls `queue_redraw` in different stages it won't be drawn multiple times?

rune-scape commented:
it does get called in multiple stages, specifically in `gui_input` and in `NOTIFICATION_INTERNAL_PHYSICS_PROCESS` and somehow in another place , sometimes, (beware i tested this on 4.2) i tested how many times per iteration it was drawing, and it was between 2 and 3 every frame when scrolling vigarously

KoBeWi commented:
Ah yeah, the queue is flushed multiple times per iteration. So it could happen e.g. if you queue redraw in both process frame and physics frame. Weird that it happens tho.

MewPurPur commented:
CodeEdit should not be using physics process, it's a temporary ugly solution to get smooth scrolling that causes bugs, but no one has managed to solve this yet.

rune-scape commented:
i see.. i.. assumed it was simply copied over from `VScroll` and never changed, im curious what the bug is, i havent noticed it yet

rune-scape commented:
sorry for not being clear about how it draws multiple times every frame but i didn't expect it to be unbelievable
i could split out out the redraw part of this PR, it seems controversial, but heres a test to show im not making this up:
[CodeEditMultipleDrawTest.zip](https://github.com/user-attachments/files/15746201/CodeEditMultipleDrawTest.zip)

i tested it just now on master on a debug linux build and it seems to draw 8! times per process iteration when scrolling vigarously

and @MewPurPur, i asked about the smooth scrolling bug bc in this PR ive changed the TextEdit to use `NOTIFICATION_INTERNAL_PROCESS` instead of physics process (it seemed like you were saying that causes a bug), and i have some people using this, i'd like to know how serious the bug is.

MewPurPur commented:
https://github.com/godotengine/godot/issues/28385

lawnjelly commented:
> it very much does draw multiple times per process loop (maybe different meaning than frame) bc queue_redraw adds a message to the MessageQueue which gets flushed multiple times per iteration

Some options off top of my head:
* Frame specific `MessageQueue`.
* Bespoke list that is flushed once per frame (i.e. `CanvasItem` redraw queue).

I've used the latter multiple times, notably for the physics interpolation. An advantage here is you'd have tighter control of the timing of the calls.

Frame specific `MessageQueue` I suspect would inevitably end up getting flushed multiple times and you'd end up with the same problem.

Another potential snag in either case is if the update caused side effects, like queuing the update of another `CanvasItem`, or adding stuff to the `MessageQueue`, or even setting up recursive loops, so you might need to limit what could be done in these updates, or deal with this possibility explicitly.

Calinou reviewed (approved):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/5833f597865c773fae3ee09fc4e31d4a243f812d), it works as expected. I've tested the script editor and didn't notice any regressions to common behavior (code folding, autocompletion, multiple carets).

## Benchmark

On average, this PR is **2.16× faster** than `master` with an optimized editor build (`optimize=speed lto=full`) on the MRP linked in the PR.

However, I get a reproducible slowdown in terms of editor FPS when opening the [Platformer 2D](https://github.com/godotengine/godot-demo-projects/tree/master/2d/platformer)'s `player.gd` script in the editor with the following editor settings:

- Script is scrolled to its first line, with the caret on the first line and first character (no text is selected).
- Output panel is cleared and collapsed.
- Window is maximized on a 4K display at 100% scaling.
- Editor started with the `--print-fps` command line argument.
- V-Sync Mode: **Disabled**.
- **Update Continuously** checked.

Before this PR, I get 861 FPS (1.16 mspf) in a maximized editor window on average, while after this PR, I get 758 FPS (1.31 mspf) on average.

I've made sure I'm CPU-bottlenecked in this situation (GPU utilization does not exceed 40% in this scenario). This issue does not occur at lower resolutions (where less of the script is visible), such as a 1920x1080 window of the editor. In this case, I'm even more CPU-bottlenecked but this PR wins out over `master`.

Either way, I figure I'd point this out as this may point out at some scalability issues in the new code (e.g. methods that have `O(N²)` scaling has replaced methods that were previously `O(N)`).

<details open>

<summary>PC specifications</summary>

- **CPU:** Intel Core i9-13900K
- **GPU:** NVIDIA GeForce RTX 4090
- **RAM:** 64 GB (2×32 GB DDR5-5800 C30)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Linux (Fedora 39)
</details>

### Before

```
10.0878296476422ms avg draw time (99 samples)
10.0065517907191ms avg draw time (198 samples)
9.94229878640737ms avg draw time (297 samples)
9.96660463737719ms avg draw time (396 samples)
9.95996937607274ms avg draw time (495 samples)
9.95242515396992ms avg draw time (594 samples)
9.94330463987408ms avg draw time (693 samples)
9.93462313305248ms avg draw time (792 samples)
9.93317573025037ms avg draw time (891 samples)
9.9287399137863ms avg draw time (990 samples)
9.92602815969588ms avg draw time (1089 samples)
```

### After

```
4.75202184734922ms avg draw time (99 samples)
4.6268870132138ms avg draw time (198 samples)
4.60077375675291ms avg draw time (297 samples)
4.57970060483374ms avg draw time (396 samples)
4.56591904765428ms avg draw time (495 samples)
4.56037465169374ms avg draw time (594 samples)
4.5561594364447ms avg draw time (693 samples)
4.55680308919964ms avg draw time (792 samples)
4.55250055032681ms avg draw time (891 samples)
4.55153975823913ms avg draw time (990 samples)
4.55042635004273ms avg draw time (1089 samples)
```

MewPurPur commented:
I'm finding a lot of instances where _draw() is called twice in 4.3 when it was only called once in 4.2.2, could that be related? I haven't set on to make a MRP yet, but someone else encountered the same:

> _draw runs for me twice too for some reason
> but only after nesting the script with the _draw in another control
> for me it was panel

kitbdev reviewed (commented):
- fixes https://github.com/godotengine/godot/issues/93036 (after its fixed)

It looks like `queue_redraw` has an issue that lets it be called multiple times in a frame, and it is not specifically related to TextEdit or CodeEdit. It should be fixed elsewhere and `queue_redraw` should continue to be used in these files instead of this `set_process_internal` workaround.
When trying your example (https://github.com/godotengine/godot/pull/92865#issuecomment-2155763912) I could only rarely get it to draw 2-3 times per frame. 


Also, when there are many lines in a script (over 1000) the line numbers jitter left and right when adding/removing a line.

Here I'm hitting backspace and enter, though the jittering doesn't look exactly the same in gif:

![godot te line number jitter](https://github.com/godotengine/godot/assets/10054226/55a13c2f-3bf4-4a3a-8b64-09a44639b608)

rune-scape commented on scene/gui/code_edit.cpp:
```diff
@@ -1253,9 +1270,9 @@ bool CodeEdit::is_drawing_executing_lines_gutter() const {
 }
 
 void CodeEdit::_main_gutter_draw_callback(int p_line, int p_gutter, const Rect2 &p_region) {
+	bool hovering = get_hovered_gutter() == Vector2i(p_line, main_gutter);
```
woww,, can't believe i missed this

rune-scape commented:
reverted back to using queue redraw and `NOTIFICATION_INTERNAL_PHYSICS_PROCESS`

> Also, when there are many lines in a script (over 1000) the line numbers jitter left and right when adding/removing a line.

i also removed the change that caused this and tested it no longer happens, i would appreciate u confirming also @kitbdev 

> When trying your example (https://github.com/godotengine/godot/pull/92865#issuecomment-2155763912) I could only rarely get it to draw 2-3 times per frame.

it was probably bc i was using the debug build of the editor, and it was hitting all 8 max physics frames per process frame, it makes the debug version of the editor alot harder to use

rune-scape commented:
> However, I get a reproducible slowdown in terms of editor FPS when opening the [Platformer 2D](https://github.com/godotengine/godot-demo-projects/tree/master/2d/platformer)'s player.gd script in the editor

> this may point out at some scalability issues in the new code (e.g. methods that have O(N²) scaling has replaced methods that were previously O(N)).

@Calinou im quite sure theres only one place this happened (replacing a hashmap lookup with `Vector.has`), and it only applies to having multiple cursors, 

when i tested this i didnt get any meaningful slowdown, (probly bc i dont have a 4k monitor)
but i have a slight suspicion that changing it to use `NOTIFICATION_INTERNAL_PROCESS` caused this, and now that its changed back to use `NOTIFICATION_INTERNAL_PHYSICS_PROCESS` it may not have that slowdown

AThousandShips commented on scene/gui/text_edit.cpp:
```diff
@@ -2927,6 +2951,35 @@ void TextEdit::_update_ime_text() {
+
+	_update_ime_window_position();
+
+	if (DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_VIRTUAL_KEYBOARD) && virtual_keyboard_enabled) {
```
```suggestion
	if (virtual_keyboard_enabled && DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_VIRTUAL_KEYBOARD)) {
```
Faster check first

rune-scape replied:
its not in the hotpath and wasnt causing problems, but ill make the change

kitbdev commented on scene/gui/code_edit.cpp:
```diff
@@ -1445,19 +1460,59 @@ bool CodeEdit::is_line_numbers_zero_padded() const {
-	tl.instantiate();
-	tl->add_string(fc, theme_cache.font, theme_cache.font_size);
-	int yofs = p_region.position.y + (get_line_height() - tl->get_size().y) / 2;
+	if (!get_rect().intersects(p_region)) {
+		return;
+	}
```
This is not correct, if the CodeEdit's rect is not at the origin then line numbers won't appear.

rune-scape replied:
i see it now, my bad

kitbdev commented on scene/gui/code_edit.cpp:
```diff
@@ -3602,14 +3663,12 @@ void CodeEdit::_text_changed() {
+	line_number_digits = new_line_number_digits;
+	_update_line_number_gutter_width();
 
 	lc = get_line_count();
```
```suggestion
```
This line is no longer needed, since `lc` doesn't change from when it is set.

kitbdev commented on scene/gui/code_edit.cpp:
```diff
@@ -1436,7 +1445,13 @@ bool CodeEdit::is_draw_line_numbers_enabled() const {
 
 void CodeEdit::set_line_numbers_zero_padded(bool p_zero_padded) {
-	p_zero_padded ? line_number_padding = "0" : line_number_padding = " ";
+	String old_line_number_padding = line_number_padding;
+	line_number_padding = p_zero_padded ? "0" : " ";
+	if (line_number_padding == old_line_number_padding) {
+		return;
+	}
```
```suggestion
	String new_line_number_padding = p_zero_padded ? "0" : " ";
	if (line_number_padding == new_line_number_padding) {
		return;
	}
	line_number_padding = new_line_number_padding;
```

This can be switched so `line_number_padding` isn't set if it is the same.

kitbdev reviewed (changes requested):
The line number jitter when there are too many lines issues is fixed.

There is an issue when deleting the longest line, the width doesn't go to the second longest line:

https://github.com/user-attachments/assets/fcacb6f7-8b64-4671-a1eb-f34e3cfffb9a

Also there is an issue with the height when there is lots of wrapped text, the scrollbar doesn't show or is too small:

![image](https://github.com/user-attachments/assets/69a6c98c-6406-4582-9409-f24ad21e442a)

rune-scape committed:
CodeEdit: improve render speed

rune-scape commented:
@kitbdev thank you for reviewing!!! everything u mentioned should be fixed now,

seems the max line height/width being invalidated by setting it to -1 was not a good idea,, lol

akien-mga commented:
Needs rebase.

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -5703,7 +5730,6 @@ Vector<String> TextEdit::get_line_wrapped_text(int p_line) const {
 /* Viewport */
 // Scrolling.
 void TextEdit::set_smooth_scroll_enabled(const bool p_enabled) {
-	v_scroll->set_smooth_scroll_enabled(p_enabled);
```
Why was this removed? This stops smooth scrolling from working when hovering over the scrollbar.

rune-scape replied:
ach, one of the things i missed when i split this PR from a bigger commit

kitbdev commented on scene/gui/text_edit.cpp:
```diff
@@ -2924,6 +2922,35 @@ void TextEdit::_update_ime_text() {
 	queue_redraw();
 }
 
+void TextEdit::_on_focus_enter() {
+	if (caret_blink_enabled) {
+		caret_blink_timer->start();
+	} else {
+		draw_caret = true;
+	}
+
+	_update_ime_window_position();
+
+	if (virtual_keyboard_enabled && DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_VIRTUAL_KEYBOARD)) {
```
```suggestion
void TextEdit::_show_virtual_keyboard() {
	if (virtual_keyboard_enabled && DisplayServer::get_singleton()->has_feature(DisplayServer::FEATURE_VIRTUAL_KEYBOARD)) {
```

This can be just about the virtual keyboard, the caret blink stuff and the ime window update can be kept back in the focus enter notification. When the mouse is released, the caret blink doesn't need to be updated and a mouse press closes the ime so it should never be up when a mouse release happens.

rune-scape replied:
i was hesitant about changing any of the actual behavior of the text editor, i wanted to keep these as strictly optimizations, im not sure if  `_update_ime_window_position` is ok to remove here? it seems important for showing the virtual keyboard
but ill refactor the caret blink stuff.

kitbdev reviewed (commented):
Previous issues fixed.

For https://github.com/godotengine/godot/issues/93036 it is fixed, See my comment here https://github.com/godotengine/godot/issues/93036#issuecomment-2263748558.
Hovering below the last line in the breakpoint gutter will now show the hovered breakpoint on the last line. This now matches the click behavior, since clicking there will add a breakpoint to the last line. If it isn't wanted it can be fixed in another PR.

Just a couple of minor things:

rune-scape commented on scene/gui/text_edit.cpp:
```diff
@@ -8394,6 +8448,8 @@ TextEdit::TextEdit(const String &p_placeholder) {
 
 	v_scroll->connect("scrolling", callable_mp(this, &TextEdit::_v_scroll_input));
 
+	connect("lines_edited_from", callable_mp(this, &TextEdit::_clear_syntax_highlighting_cache).unbind(2));
```
i this way of connecting it to the signal is easier to follow and if there is a new function added that edits lines, calling `_clear_syntax_highlighting_cache` doesn't have to be considered, less burden on a programmer adding functions

kitbdev replied:
I think its more clear if it is directly called rather than through a signal, and there shouldn't be anything that changes about editing lines that doesn't go through `_clear()` or `_text_changed()`, since both `_base_remove_text` and `_base_insert_text` call `_text_changed`.
Also if `set_block_signals` is used, it shouldn't disrupt something internal like clearing the syntax highlighting cache.

rune-scape replied:
i understand, that makes sense
i had no idea about `set_block_signals()` or `_text_changed()` i see now

kitbdev reviewed (approved):
Haven't tested the performance, but the functionality looks good to me.

rune-scape commented on scene/gui/text_edit.h:
```diff
@@ -568,8 +564,10 @@ class TextEdit : public Control {
 
 	/* Syntax highlighting. */
 	Ref<SyntaxHighlighter> syntax_highlighter;
+	HashMap<int, Vector<Pair<int64_t, Color>>> syntax_highlighting_cache;
```
it does, but this was the biggest speedup and the first improvement i made
its the reason the difference between the times for 'with syntax highlighting' and 'without syntax highlighting' get so much closer after this PR
it was a combo of multiple hashmap lookups:
- linear search of dict keys using `Dictionary::next()` uses a hash lookup to find the key to get the next key
- if its the right key, then that key is used to find the color value
- the color value is a dictionary and the actual color has to be gotten with a hashmap lookup
also converting to and from variants, (a surprisingly large amount of cpu time)

all of this is done for **every character in every line on the screen and in the minimap!!! every frame!**
very hot code

i replaced it with a simple linear search through a packed array

akien-mga commented:
Thanks!

arkology commented:
I have tested speed improvement by changing zoom inside script editor and difference is really noticeable🚀
